### PR TITLE
Block processing electra

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

https://github.com/sigp/lighthouse/issues/5605

## Tasks

- [x] [Withdrawals](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#withdrawals)
  - [x] [Updated get_expected_withdrawals](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#updated-get_expected_withdrawals)
  - [x] [Updated process_withdrawals](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#updated-process_withdrawals)
- [x] [Execution payload](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#execution-payload)
  - [x] [Modified process_execution_payload](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#modified-process_execution_payload)
- [x] [Operations](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#operations)
  - [x]         [Attestations](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#attestations)
    - [x]             [Modified process_attestation](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#modified-process_attestation)
  - [x]         [Deposits](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#deposits)
    - [x]             [Updated apply_deposit](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#updated--apply_deposit)
    - [x]             [New is_valid_deposit_signature](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#new-is_valid_deposit_signature)
    - [x]             [Modified add_validator_to_registry](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#modified-add_validator_to_registry)
    - [x]             [Updated get_validator_from_deposit](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#updated-get_validator_from_deposit)
  - [x]         [Voluntary exits](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#voluntary-exits)
    - [x]             [Updated process_voluntary_exit](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#updated-process_voluntary_exit)
  - [x]         [Execution layer withdrawal requests](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#execution-layer-withdrawal-requests)
    - [x]             [New process_execution_layer_withdrawal_request](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#new-process_execution_layer_withdrawal_request)
  -   [x]         [Deposit receipts](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#deposit-receipts)
      - [x]             [New process_deposit_receipt](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#new-process_deposit_receipt)
    - [x]         [Consolidations](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#consolidations)
      - [x]             [New process_consolidation](https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md#new-process_consolidation)

### CODE REVIEW

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 1]
```diff
@@ -36,6 +39,18 @@ pub fn process_operations<E: EthSpec, Payload: AbstractExecPayload<E>>(
         process_bls_to_execution_changes(state, bls_to_execution_changes, verify_signatures, spec)?;
     }
 
+    if state.fork_name_unchecked() >= ForkName::Electra {
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/process_operations.rs]: `fork.electra_enabled()`

File: consensus/state_processing/src/per_block_processing/errors.rs [hunk 2]
```diff
@@ -89,6 +89,46 @@ pub enum BlockProcessingError {
...
+    ConsolidationChurnLimitTooLow {
+        churn_limit: u64,
+        minimum: u64,
+    },
+    MatchingSourceTargetConsolidation {
+        index: u64,
+    },
+    InactiveConsolidationSource {
+        index: u64,
+        current_epoch: Epoch,
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/errors.rs:105]: Is this error property useful? This error will appear in the context of a block at slot N with root X. It's trivial to get the epoch from there

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 3]
```diff
@@ -447,6 +515,280 @@ pub fn apply_deposit<E: EthSpec>(
...
+        // Verify the validator has been active long enough
+        if state.current_epoch()
+            < validator
+                .activation_epoch
+                .safe_add(spec.shard_committee_period)?
+        {
+            continue;
+        }
+
+        let pending_balance_to_withdraw = state.get_pending_balance_to_withdraw(index)?;
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/process_operations.rs:585]: `get_pending_balance_to_withdraw` iterates over the full list of withdrawals making it a O(n^2) operation technically. We process only a few withdrawal requests per block, but we should watch to ensure we don't need a cache.

**Review by dapplion — APPROVED** [review]
All logic is correct with respect to the old spec, good to merge to me if we prefer to update to latest spec latter

File: consensus/state_processing/src/per_block_processing/errors.rs (hunk 2 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/errors.rs:105]: this error is actually not necessary in the latest spec, mark's already got it removed here: https://github.com/sigp/lighthouse/pull/5994/files

so I'll just leave it

File: consensus/state_processing/src/per_block_processing/process_operations.rs (hunk 3 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/process_operations.rs:585]: maybe we should add per-operations metrics in a separate pr?

File: consensus/state_processing/src/per_block_processing/process_operations.rs (hunk 1 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/process_operations.rs]: added here 7c17098b20b07bcc5491db3c2f63c14c9ccbe0b8

File: consensus/state_processing/src/per_block_processing/process_operations.rs (hunk 3 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/process_operations.rs:585]: https://github.com/sigp/lighthouse/issues/5999

**Comment by realbigsean** [comment]
@mergify queue

**Pull request merged**